### PR TITLE
SQL: Implement TOP as an alternative to LIMIT

### PR DESCRIPTION
**Pull request by matriv**
Add basic support for `TOP X` as a synonym to LIMIT X which is used
by [MS-SQL server](https://docs.microsoft.com/en-us/sql/t-sql/queries/top-transact-sql?view=sql-server-ver15),
e.g.:

```
SELECT TOP 5 a, b, c FROM test
```

TOP in SQL server also supports the `PERCENTAGE` amd `WITH TIES`
keywords which this implementation doesn't.

Don't allow usage of both TOP and LIMIT in the same query.

Refers to #41195

### CODE REVIEW

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/LogicalPlanBuilder.java [hunk 1]
```diff
@@ -243,4 +258,8 @@ public LogicalPlan visitTableName(TableNameContext ctx) {
         TableIdentifier tableIdentifier = visitTableIdentifier(ctx.tableIdentifier());
         return new UnresolvedRelation(source(ctx), tableIdentifier, alias, ctx.FROZEN() != null);
     }
+
+    private Limit wrapWithLimit(LogicalPlan plan, Source source, Token limit) {
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/LogicalPlanBuilder.java:262]: I'd prefer calling the method simply `limit` - it's not really wrapping but instead adding a new node to the tree.

**Review by costin — APPROVED** [review]
LGTM

File: docs/reference/sql/language/syntax/commands/select.asciidoc [hunk 2]
```diff
@@ -371,26 +371,52 @@ all are equally relevant.
...
+The `TOP` clause restricts (limits) the number of rows returned using the format:
+
+[source, sql]
+----
+SELECT TOP <count> <list of fields> ...
+----
+
+where
+
+count:: is a positive integer or zero indicating the maximum *possible* number of results being returned (as there might be less matches
```
Comment (astefan) [inline docs/reference/sql/language/syntax/commands/select.asciidoc:409]: I think `fewer` goes better with `matches`.

File: docs/reference/sql/language/syntax/commands/select.asciidoc [hunk 3]
```diff
@@ -371,26 +371,52 @@ all are equally relevant.
...
+than the limit). If `0` is specified, no results are returned.
+
+[source, sql]
+----
+include-tagged::{sql-specs}/docs/docs.csv-spec[topBasic]
+----
+
+[NOTE]
+<<sql-syntax-top, `TOP`>> and <<sql-syntax-limit, `LIMIT`>> cannot be used in combination in the same query and error is returned.
```
Comment (astefan) [inline docs/reference/sql/language/syntax/commands/select.asciidoc:418]: `...cannot be used together in the same query and an error is returned otherwise` sounds better? If so, same sentence was used for `LIMIT` and the change is needed there as well.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/LogicalPlanBuilder.java [hunk 4]
```diff
@@ -155,6 +158,15 @@ else if (!selectTarget.isEmpty()) {
         if (ctx.setQuantifier() != null && ctx.setQuantifier().DISTINCT() != null) {
             query = new Distinct(source(ctx.setQuantifier()), query);
         }
+
+        // TOP
+        SqlBaseParser.TopClauseContext topClauseContext = ctx.topClause();
+        if (topClauseContext != null) {
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/LogicalPlanBuilder.java:164]: Why not combining all three conditions in the same `if`?  More readable this way?

**Review by astefan — APPROVED** [review]
LGTM. Left few minor comments.

File: docs/reference/sql/language/syntax/commands/select.asciidoc (hunk 2 above)
Comment (matriv) [reply docs/reference/sql/language/syntax/commands/select.asciidoc:409]: Was a copy from LIMIT, so fixing it there as well.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/LogicalPlanBuilder.java [hunk 5]
```diff
@@ -105,8 +105,11 @@ public LogicalPlan visitQueryNoWith(QueryNoWithContext ctx) {
         if (limitClause != null) {
             Token limit = limitClause.limit;
             if (limit != null && limitClause.INTEGER_VALUE() != null) {
-                plan = new Limit(source(limitClause), new Literal(source(limitClause),
-                        Integer.parseInt(limit.getText()), DataTypes.INTEGER), plan);
+                if (plan instanceof Limit) {
+                    throw new ParsingException(source(limitClause), "Cannot use both TOP and LIMIT in the same query");
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/LogicalPlanBuilder.java:109]: I would rephrase the error a bit - `TOP and LIMIT are not allowed in the same query - use one or the other`.
